typo: fix missing backtick (#20202)
cloudflare/cloudflare-docs: src/content/docs/agents/concepts/what-are-agents.mdx

---
title: Agents
pcx_content_type: concept
sidebar:
  order: 2

---

import { Render } from "~/components";

> ### What are agents? 

Reviewer: ```suggestion
### What are agents?

```

> An agent is an AI system that can autonomously execute tasks by making decisions about tool usage and process flow. Unlike traditional automation that follows predefined paths, agents can dynamically adapt their approach based on context and intermediate results. Agents are also distinct from co-pilots (e.g. traditional chat applications) in that they can fully automate a task, as opposed to simply augmenting and extending human input. 

Reviewer: ```suggestion
An agent is an AI system that can autonomously execute tasks by making decisions about tool usage and process flow. Unlike traditional automation that follows predefined paths, agents can dynamically adapt their approach based on context and intermediate results. Agents are also distinct from co-pilots (for example, traditional chat applications) in that they can fully automate a task, as opposed to simply augmenting and extending human input. 
```

> - **Agents** → non-linear, non-deterministic (can change from run to run)
> - **Workflows** → linear, deterministic execution paths
> - **Co-pilots** → augmentative AI assistance requiring human intervention

Reviewer: ```suggestion
- **Agents**: non-linear, non-deterministic (can change from run to run)
- **Workflows**: linear, deterministic execution paths
- **Co-pilots**: augmentative AI assistance requiring human intervention
```
Suggestion: Typically we use colons here.

> ### Example: Booking Vacations

Reviewer: ```suggestion
### Example: Booking vacations

```

> If this is your first time working with, or interacting with agents, this example will illustrate how an agent works within a context like booking a vacation. If you're already familiar with the topic, read on. 

Reviewer: ```suggestion
If this is your first time working with, or interacting with agents, this example will illustrate how an agent works within a context like booking a vacation. If you are already familiar with the topic, read on. 
```

Imagine you're trying to book a vacation. You need to research flights, find hotels, check restaurant reviews, and keep track of your budget. 

> ##### Traditional Workflow Automation

Reviewer: ```suggestion
##### Traditional workflow automation

```

A traditional automation system follows a predetermined sequence:
- Takes specific inputs (dates, location, budget)
- Calls predefined API endpoints in a fixed order
- Returns results based on hardcoded criteria
- Cannot adapt if unexpected situations arise

![Traditional workflow automation diagram](~/assets/images/agents/workflow-automation.svg)

> ##### AI Co-pilot

Reviewer: ```suggestion
#### AI Co-pilot
```

A co-pilot acts as an intelligent assistant that:

- Provides hotel and itinerary recommendations based on your preferences
- Can understand and respond to natural language queries
- Offers guidance and suggestions
- Requires human decision-making and action for execution

![A co-pilot diagram](~/assets/images/agents/co-pilot.svg)

##### Agent

> An agent combines AI's ability to make judgements and call the relevant tools to execute the task. An agent's output will be nondeterministic given:

Reviewer: ```suggestion
An agent uses AI's ability to make judgements and calls the relevant tools to execute the task. An agent's output will be nondeterministic given:
```

- Real-time availability and pricing changes
- Dynamic prioritization of constraints
- Ability to recover from failures
- Adaptive decision-making based on intermediate results

![An agent diagram](~/assets/images/agents/agent-workflow.svg)

An agents can dynamically generate and itinerary and execute on booking reservations, similarly to what you would expect from a travel agent. 

### Three primary components of agent systems:

- **Decision Engine**: Usually an LLM (Large Language Model) that determines action steps
- **Tool Integration**: APIs, functions, and services the agent can utilize
- **Memory System**: Maintains context and tracks task progress

#### How Agents Work

> Agents operate in a continuous loop of:

Reviewer: ```suggestion
Agents operate in a continuous loop of:

```
